src/lib/pmd/PmdEngine.ts

```diff
@@ -162,34 +196,81 @@ abstract class BasePmdEngine implements RuleEngine {
 		}
 	}
 
+	/**
+	 * Indicates whether the scanner should consider a given node as representing PMD violations. Some nodes don't technically
+	 * contain violations, but should be treated as fake violations of fake rules.
+	 * @param node - A top-level node from PMD's results.
+	 * @returns {boolean} true if the node represents rule violations in the scanner's opinion.
+	 * @private
+	 */
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
+	private nodeRepresentsViolations(node: any): boolean {
+		// `file` nodes always contain violations.
+		if (node.name === 'file') {
+			return true;
+		}
+		// Other types of nodes don't technically contain violations, but sometimes they indicate problems that we want
+		// to surface in the same manner as true violations, instead of errors or warnings.
+		return HARDCODED_RULE_DETAILS.some((detail: HardcodedRuleDetail): boolean => detail.nodeIdentifier(node));
+	}
+
 	// eslint-disable-next-line @typescript-eslint/no-explicit-any
 	protected xmlToRuleResults(elements: any): RuleResult[] {
 		// Provide results for nodes that are files.
-		const files = elements.filter(e => 'file' === e.name);
-
-		return files.map(
-			(f): RuleResult => {
-				return {
-					engine: this.getName(),
-					fileName: f.attributes['name'],
-					violations: f.elements.map(
-						(v: PmdViolation) => {
-							return {
-								line: v.attributes.beginline,
-								column: v.attributes.begincolumn,
-								endLine: v.attributes.endline,
-								endColumn: v.attributes.endcolumn,
-								severity: v.attributes.priority,
-								ruleName: v.attributes.rule,
-								category: v.attributes.ruleset,
-								url: v.attributes.externalInfoUrl,
-								message: this.toText(v)
-							};
-						}
-					)
-				};
+		const violationNodes = elements.filter(this.nodeRepresentsViolations);
+
+		return violationNodes.map((node): RuleResult => {
+			switch (node.name) {
+				case 'file':
+					return this.fileNodeToRuleResult(node);
+				case 'error':
+					return this.errorNodeToRuleResult(node);
+				default:
+					// This shouldn't be possible, but it's best practice for every switch to have a default.
+					throw new SfdxError(`Encountered violation node of unexpected type ${node.name}`);
```

Comment:
If we see some data that we don't recognize at all, I feel a safer approach would be to maintain the current behavior of throwing a Warning and moving on. This way, we won't block the user for something they can't fix.

Reply:
That still happens in `emitErrorsAndWarnings()`.